dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/attribute/AttributeValueDeletionHandler.java

```diff
@@ -0,0 +1,78 @@
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import com.google.common.collect.Lists;
+import org.hisp.dhis.common.IdentifiableObject;
+import org.hisp.dhis.common.IdentifiableObjectManager;
+import org.hisp.dhis.system.deletion.DeletionHandler;
+import org.springframework.stereotype.Component;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+@Component( "org.hisp.dhis.attribute.AttributeValueDeletionHandler" )
+public class AttributeValueDeletionHandler
+    extends DeletionHandler
+{
+    private final IdentifiableObjectManager identifiableObjectManager;
+
+    private String supportedClassName;
+
+    public AttributeValueDeletionHandler( IdentifiableObjectManager identifiableObjectManager )
+    {
+        checkNotNull( identifiableObjectManager );
+
+        this.identifiableObjectManager = identifiableObjectManager;
+    }
+
+    // -------------------------------------------------------------------------
+    // DeletionHandler implementation
+    // -------------------------------------------------------------------------
+
+    @Override
+    public String getClassName()
+    {
+        return supportedClassName  + "." +  AttributeValue.class.getSimpleName();
+    }
+
+    @Override
+    public String allowDeleteAttribute( Attribute attribute )
+    {
+        for (  Class<? extends IdentifiableObject> supportedClass : attribute.getSupportedClasses() )
+        {
+            if ( identifiableObjectManager.getAllValuesByAttributes( supportedClass, Lists.newArrayList( attribute ) ).size() > 0 )
```

Comment:
Can we do a `count` or an `exists` method and use that instead @vietnguyen ? Getting all attribute values into memory is expensive and might lead to out-of-mem, and unnecessary when we just want to check whether at least one exists.

Reply:
just added the function countAllValuesByAttributes() in HibernateGenericStore to solve this

Reply:
Very good Viet, thanks. The code looks great. Approved it.